File: src/myroboticarm_firmware/myroboticarm_firmware/simple_serial_transmitter.py
#!/usr/bin/env python3

import rclpy 
from rclpy.node import Node
from std_msgs.msg import String
import serial

class SimpleSerialTransmitter(Node):
    def __init__(self):
        super().__init__("simple_serial_transmitter")

        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baud_rate", 115200)

        self.port = self.get_parameter("port").value
        self.baud_rate = self.get_parameter("baud_rate").value

        self.sub = self.create_subscription(String, "serial_transmitter", self.msg_callback, 10)
        self.arduino = serial.Serial(port= self.port , baudrate= self.baud_rate, timeout= 0.1) #Used to send serial messages to the arduino 


    def msg_callback(self, msg):  # This is activated when receiving a new message in the serial transmitter topic, now we can send the received message to the serial object using self.arduino
        self.get_logger().info("New message received, publishing on serial: %s" % self.arduino.name)
        self.arduino.write(msg.data.encode("utf-8"))

def main(args= None):
    rclpy.init(args=args)
    simple_serial_transmitter = SimpleSerialTransmitter()
    rclpy.spin(simple_serial_transmitter)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/myroboticarm_firmware/src/simple_serial_transmitter.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <libserial/SerialPort.h>

using std::placeholders::_1;

// Define the SimpleSerialTransmitter class to communicate with the Arduino over serial
class SimpleSerialTransmitter : public rclcpp::Node
{
public:
  SimpleSerialTransmitter() : Node("simple_serial_transmitter")
  {
    declare_parameter<std::string>("port", "/dev/ttyUSB0");
    std::string port = get_parameter("port").as_string();
    sub = create_subscription<std_msgs::msg::String>("chatter", 10, std::bind(&SimpleSerialTransmitter::msgcallback, this, _1));
    arduino.Open(port);
    arduino.SetBaudRate(LibSerial::BaudRate::BAUD_115200);  // Changed from Libserial to LibSerial
  }
  
  // The class simpleserialtransmitter is open for communication, but now we also need to close the connection
  ~SimpleSerialTransmitter()
  {
    arduino.Close();
  }
  
  void msgcallback(const std_msgs::msg::String &msg)
  {
    RCLCPP_INFO_STREAM(get_logger(), "New message received, publishing on serial port: " << msg.data.c_str());
    arduino.Write(msg.data); // Takes as input the data that we want to publish on the serial port
  }

private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub;
  LibSerial::SerialPort arduino;  // Changed from Libserial to LibSerial
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSerialTransmitter>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}